src/main/java/org/spongepowered/server/launch/VanillaServerMain.java

```diff
@@ -138,63 +145,137 @@ public static void main(String[] args) throws Exception {
+                    versionManifestRemote = obj.get("url").asString();
+                    break;
+                }
+            }
+
+            if (versionManifestRemote == null) {
+                throw new RuntimeException("Could not find " + MINECRAFT_SERVER_VERSION + "'s manifest URL");
+            }
+
+            JsonValue versionManifest = downloadJson(versionManifestRemote);
+            JsonObject serverObj = versionManifest.asObject()
+                    .get("downloads").asObject()
+                    .get("server").asObject();
+
+            // Find the server URL and SHA-1 digest
+            String serverRemote = serverObj.get("url").asString();
+            String sha1 = serverObj.get("sha1").asString();
+
+            downloadAndVerify(serverRemote, path, sha1);
         }
 
-        // Make sure Launchwrapper is available, or download it otherwise
         path = base.resolve(LAUNCHWRAPPER_LOCAL);
-        return Files.exists(path) || (autoDownload && downloadVerified(LAUNCHWRAPPER_REMOTE, path));
+
+        if (!Files.exists(path)) {
+            if (!autoDownload) {
+                return false;
+            }
+
+            // Make sure Launchwrapper is available, or download it otherwise
+            download(LAUNCHWRAPPER_REMOTE, path);
+        }
+
+        return true;
     }
 
+    private static JsonValue downloadJson(String remote) throws IOException {
+        URLConnection con = new URL(remote).openConnection();
+        JsonValue json;
+
+        try (BufferedInputStream source = new BufferedInputStream(con.getInputStream())) {
+            json = Json.parse(new InputStreamReader(source, StandardCharsets.UTF_8));
+        }
 
-    private static boolean downloadVerified(String remote, Path path) throws IOException, NoSuchAlgorithmException {
+        return json;
+    }
+
+    /**
+     * Downloads a file and verify its digest.
+     *
+     * @param remote The file URL
+     * @param path The local path
+     * @param sha1 The SHA-1 expected digest
+     * @throws IOException If there is a problem while downloading the file
+     */
+    private static void downloadAndVerify(String remote, Path path, String sha1) throws IOException {
         Files.createDirectories(path.getParent());
 
         String name = path.getFileName().toString();
         URL url = new URL(remote);
 
         System.out.println("Downloading " + name + "... This can take a while.");
         System.out.println(url);
+
         URLConnection con = url.openConnection();
-        MessageDigest md5 = MessageDigest.getInstance("MD5");
+        BufferedInputStream stream = new BufferedInputStream(con.getInputStream());
+        MessageDigest digest;
+
+        try {
+            digest = MessageDigest.getInstance("SHA-1");
+        } catch (NoSuchAlgorithmException e) {
+            // This cannot happen because JVM is required to support SHA-1
+            throw new RuntimeException(e);
```

Comment:
I think it's cleaner if you continue propagating the `NoSuchAlgorithmException` by adding `throws NoSuchAlgorithmException` here. That way it's captured by the `main` method and doesn't need to be handled specifically.